src/aspire/basis/polar_2d.py

```diff
@@ -0,0 +1,111 @@
+import logging
+import numpy as np
+import finufftpy
+
+from aspire.utils import ensure
+from aspire.utils.matrix import roll_dim, unroll_dim
+from aspire.utils.matlab_compat import m_reshape
+from aspire.basis import Basis
+
+logger = logging.getLogger(__name__)
+
+
+class PolarBasis2D(Basis):
+    """
+    Define a derived class for polar Fourier expansion for 2D images
+    """
+
+    def __init__(self, size, nrad=None, ntheta=None):
+        """
+        Initialize an object for the 2D polar Fourier basis class
+
+        :param size: The size of the vectors for which to define the basis.
+            Currently only square images are supported.
+        :param nrad: The number of points in the radial dimension.
+        :param ntheta: The number of points in the angular dimension.
+        """
+
+        ndim = len(size)
+        ensure(ndim == 2, 'Only two-dimensional basis functions are supported.')
+        ensure(len(set(size)) == 1, 'Only square domains are supported.')
+
+        self.nrad = nrad
+        if nrad is None:
+            self.nrad = self.nres // 2
+
+        self.ntheta = ntheta
+        if ntheta is None:
+            # try to use the same number as Fast FB basis
+            self.ntheta = 8 * self.nrad
+
+        super().__init__(size)
+
+    def _build(self):
+        """
+        Build the internal data structure to 2D polar Fourier basis
+        """
+        logger.info('Expanding 2D image in a polar Fourier basis')
+
+        self.count = self.nrad * self.ntheta
+        self.sz_prod = self.sz[0] * self.sz[1]
+
+        # precompute the basis functions in 2D grids
+        self.freqs = self._precomp()
+
+    def _precomp(self):
+        """
+        Precomute the basis functions on a polar Fourier grid
+        """
+        omega0 = 2 * np.pi / (2 * self.nrad - 1)
+        dtheta = 2 * np.pi / self.ntheta
+
+        # Note: The size of ntheta can be reduced by half if only evaluate_t function is needed as
+        # in the Matlab implementation, np.zeros((2, self.nrad * self.ntheta//2)).
+        freqs = np.zeros((2, self.nrad * self.ntheta))
+        for i in range(self.ntheta):
+            freqs[0, i * self.nrad: (i + 1) * self.nrad] = np.arange(self.nrad) * np.sin(i * dtheta)
+            freqs[1, i * self.nrad: (i + 1) * self.nrad] = np.arange(self.nrad) * np.cos(i * dtheta)
+
+        freqs *= omega0
+        return freqs
+
+    def evaluate_t(self, x):
```

Comment:
Why don't we have `evaluate`? It should be fairly straightforward to compute the dual of `evaluate_t`.

Reply:
Yes. I thought this before. There is no implementation of evaluate in the original Matlab code. It will be great if I can be shown the exact equation with the right normalization factor.

Reply:
Done.